# Laboratorio de regresión - 3

## Significancia de factores

|                |   |
:----------------|---|
| **Nombre**     | Jesús Alejandro García García  |
| **Fecha**      |  04-09-25 |
| **Expediente** |  751749 |

Descarga el archivo de publicidad y carga los datos (Advertising.csv).

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score
import seaborn as sns
from scipy import stats

In [2]:
df=pd.read_csv('Advertising.csv')
df.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [3]:
df.drop('Unnamed: 0',axis=1, inplace=True)
df.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   TV         200 non-null    float64
 1   radio      200 non-null    float64
 2   newspaper  200 non-null    float64
 3   sales      200 non-null    float64
dtypes: float64(4)
memory usage: 6.4 KB


**¿Hay alguna relación entre el presupuesto para publicidad y las ventas?**

Nuestra primera meta debe ser determinar si hay evidencia en los datos de que haya una asociación entre estas variables.

- ¿Por qué? ¿Qué resultaría si nos diéramos cuenta de la falta de relación entre el presupuesto de publicidad y las ventas?

Si no hubiera una relación entre las variables, no podríamos realizar una regresión, ya que una no tiene el efecto sobre la otra.

**¿Qué tan fuerte es esta relación?**
Asumiendo que existe esta relación, ¿nos sirve conocer el impacto que tiene invertir en publicidad en las ventas?

La publicidad y las ventas tienden a tener un alto nivel de correlacion. Esto nos puede ayudar a conocer el impacto que la publicidad tiene sobre las ventas.

**¿Cuáles medios están asociados con las ventas? ¿Qué tan grande es la asociación entre un medio específico y las ventas?**

Hay 3 medios distintos en los datos. ¿Sirve invertir en los 3? ¿Conviene más invertir sólo en uno?

**¿Qué tan seguros estamos de que podríamos predecir ventas futuras?**

**¿La relación es lineal?**

**¿Hay sinergia entre estos medios?**

Puede ser que gastar \\$50,000 en publicidad y otros \\$50,000 en radio es mejor opción que gastar \\$100,000 en televisión. A esto le llamamos efecto de interacción.

Podemos usar regresión lineal para responder todas estas preguntas.

Realiza una regresión lineal:

$$ \text{ventas} \approx \beta_0 + (\beta_1)(\text{TV})$$

In [5]:
x1=df.TV.values.reshape([-1, 1])
x2=df.radio.values.reshape([-1,1])
x3=df.newspaper.values.reshape([-1,1])
y=df.sales
X=np.hstack([x1])
lr=LinearRegression()
lr.fit(X,y)
lr.intercept_

7.032593549127693

In [6]:
lr.coef_

array([0.04753664])

In [7]:
r2_score(y,lr.predict(X))

0.611875050850071

### Verificando la precisión de nuestros coeficientes estimados

Recuerda que en el mundo real hay ruidos y errores de medición. Siempre se asume que la verdadera relación entre $X$ y $Y$ es $$Y = \beta_0 + \beta_1 X + \epsilon$$

Se asume que el término de error es independiente de $X$ (el error siempre es el mismo sin importar el valor de $X$). Este modelo describe a la *línea de regresión de la población*, que es la mejor aproximación de la verdadera relación entre $X$ y $Y$. Cuando usamos mínimos cuadrados encontramos la *línea de mínimos cuadrados*.

¿Cuál es la diferencia entre población y muestra?

La muestra es un subconjunto de la población, la cual nos puede ayudar a realizar este tipo de experimentos. La población son todos los individuos. 

¿Cuál crees que sea la diferencia entre hacer una regresión con todos los datos de la población y una muestra de ella?

La mayor diferencia sería que la regresión con la muestra solo nos puede dar aproximados, ya que la muestra no contiene todos los datos. Mientras que realizar la regresión con los datos sería un resultado preciso.

La línea de regresión de la población no se puede observar. El concepto de comparar estas líneas es una extensión natural del acercamiento estadístico estándar de usar información de una muestra para estimar características de una población grande.

Imagina que quieres encontrar la altura promedio de un mexicano $\mu$. Medir a todos y cada uno de los mexicanos en situaciones similares, con la misma regla, mismo operador, y otras incontables formas de minimizar la variación de la medida es una tarea imposible. Lo que podemos asumir es que $\hat{\mu} = \bar{y}$. La media poblacional y la media muestral son diferentes, pero la media muestral es usualmente un buen estimado.

De la misma manera, como no contamos con el 100% de la información para hacer una regresión, los coeficientes $\beta_0$ y $\beta_1$ son desconocidos. Podemos estimarlos usando mínimos cuadrados, encontrando $\hat{\beta_0}$ y $\hat{\beta_1}$. Puede que las muestras que tengamos en ese momento estén un poco por encima de la media, pero otras muestras en otro momento puede que estén debajo de la media. En general, esperamos que el promedio de las aproximaciones $\hat{\mu}$ aproxime a $\mu$.

Esto lleva a la pregunta: ¿qué tan cercanos son nuestros coeficientes estimados a los verdaderos coeficientes? Utilizamos el concepto de error estándar para evaluar esto.

$$ \text{Var}(\hat{\mu})=\text{SE}(\hat{\mu})^2 = \frac{\sigma^2}{n} $$

Donde $\sigma$ es la desviación estándar de cada una de las observaciones $y_i$ de $Y$. El error estándar nos dice la cantidad promedio que el estimado difiere del valor verdadero. Podemos ver en la fórmula que entre más observaciones tengamos el error se hace más pequeño. Las fórmulas para errores estándar de $\hat{\beta_0}$ y $\hat{\beta_1}$ son:

$$ \text{SE}(\hat{\beta_0})^2 = \sigma^2 [\frac{1}{n} + \frac{\bar{x}^2}{\sum_{i=1}^n (x_i - \bar{x})^2}]$$

$$ \text{SE}(\hat{\beta_1})^2 = \frac{\sigma^2}{\sum_{i=1}^n (x_i - \bar{x})^2}$$

$$ \sigma^2 = \text{Var}(\epsilon) = \text{RSE}^2 = \frac{\text{RSS}}{n-p}$$

Para que estas fórmulas sean validas asumimos que los errores $\epsilon_i$ tienen varianza común $\sigma^2$ y que no están correlacionados.

Calcula los errores estándar de los coeficientes

In [8]:
x_media=np.mean(x1)
y_media=np.mean(y)
n=len(y)
p=2
RSS=sum((lr.predict(x1)-y)**2)
sigma2=RSS/(n-p)

In [9]:
RSS
sigma2

10.618841328946218

In [10]:
SE_Bo2=sigma2*(1/n + ((x_media**2)/sum((x1-x_media)**2)))
SE_Bo=np.sqrt(SE_Bo2)
SE_Bo

array([0.45784294])

In [11]:
SE_B12=sigma2/sum((x1-x_media)**2)
SE_B1=np.sqrt(SE_B12)
SE_B1

array([0.00269061])

Estos errores se pueden usar para calcular intervalos de confianza. Un intervalo de confianza del $95\%$ se define como un rango de valores en el cuál se encuentra el desconocido valor verdadero con un $95\%$ de probabilidad.

Otra forma de verlo es que si tomamos muestras repetidas y construimos un intervalo de confianza para cada una, el $95\%$ de los intervalos creados van a contener el valor verdadero. Para la regresión el intervalo de confianza del $95\%$ toma la forma:

$$ \hat{\beta_j} \pm 2\text{SE}(\hat{\beta_j})$$

Calcula los intervalos de confianza para los coeficientes estimados:

In [12]:
bo_bajo=lr.intercept_-2*SE_Bo
bo_alto=lr.intercept_+2*SE_Bo
b1_bajo=lr.coef_-2*SE_B1
b1_alto=lr.coef_+2*SE_B1

In [13]:
print(bo_bajo,bo_alto,b1_bajo,b1_alto)


[6.11690767] [7.94827943] [0.04215543] [0.05291785]


Los errores estándar también se usan para realizar pruebas de hipótesis. La prueba de hipótesis más común es probar la hipótesis nula de:

$$ H_0: \text{No hay relación entre } X \text{ y } Y \ \ \ \ (\beta_1=0)$$

contra la hipótesis alternativa:
$$ H_0: \text{Hay alguna relación entre } X \text{ y } Y \ \ \ (\beta_1 \neq 0)$$

Explica con tus palabras el significado de la hipótesis nula y la hipótesis alternativa.

La hipótesis alternativa es la predicción que estás realizando. La hipótesis nula es la que indica que no existe una diferencia.

Para probal la hipótesis nula debemos determinar si nuestro estimado $\hat{\beta_1}$ de $\beta_1$ está lo suficientemente alejado de cero para que podamos decir con confianza que este valor no es cero. 

¿Qué tan lejos? Depende de qué tanta confianza tengamos en el estimado encontrado. Si nuestro error estándar es pequeño y nuestro estimado está alejado de cero podríamos decir que hay muy poca probabilidad de que el valor verdadero sea 0. En cambio, si nuestro error estándar es grande y nuestro estimado está muy cerca de cero, entonces podrías ser que el valor verdadero sea cero y que no haya relación entre las variables.

Se calcula un *estadístico t* dado por
$$ t = \frac{\hat{\beta_j} - \mu}{\text{SE}(\hat{\beta_j})} $$

donde $\mu$ es el valor contra el que queremos probar.

Calcula el estadístico t para tus coeficientes estimados, usando como referencia la prueba de hipótesis.

In [14]:
t_bo=lr.intercept_/SE_Bo
t_bo

array([15.36027517])

In [15]:
t_b1=lr.coef_/SE_B1
t_b1

array([17.6676256])

La distribución t tiene forma de campana y se parece bastante a la distribución normal cuando $n > 30$. Ya sólo es cuestión de calcular la probabilidad de observar cualquier número tal que su valor absoluto sea igual o mayor que el valor absoluto del estadístico t calculado. En otras palabras:
$$ P(|x| \geq |t|) $$

A esta probabilidad la llamamos *p-value*. Un *p-value* pequeño indica que es poco probable que exista por puro azar una relación significativa entre predictor y respuesta, en caso de que no haya una asociación real entre predictor y respuesta. En otras palabras, el *p-value* te dice la probabilidad de que parezca que hay relación cuando no la hay.

Si el *p-value* es pequeño, inferimos que sí hay una asociación entre el predictor y la respuesta, y **rechazamos la hipótesis nula**.
  

¿Qué tan pequeño? Depende de la aplicación. Un valor muy común es del $5\%$.

Utiliza el siguiente código para calcular el *p-value* para tus coeficientes

`from scipy import stats`

`p_bj = 2*(1 - stats.t.cdf(np.abs(t_bj), n-p))`

In [16]:
p_b0 = 2*(1 - stats.t.cdf(np.abs(t_bo), n-p))
print(f'El p-value para b0 es: {p_b0}')

El p-value para b0 es: [0.]


In [17]:
p_b1 = 2*(1 - stats.t.cdf(np.abs(t_b1), n-p))
print(f'El p-value para b1 es: {p_b1}')

El p-value para b1 es: [0.]


¿Se rechaza la hipótesis nula? ¿Qué significa?

Decidimos rechazar la hipótesis nula, ya que los p-values de los coeficientes son muy pequeños, lo que nos indica que sí existe una relación entre las ventas y la publicidad en la TV.

Realiza otras dos regresiones. Ya tienes hecha la regresión de ventas dado el gasto en publicidad de TV. Realiza la regresión para gastos en radio y gastos en periódico. Organiza las respuestas para que debajo de esta celda se tenga:
- Título de regresión
- Coeficientes estimados
- Errores estándar de los coeficientes
- Intervalos de confianza
- Estadísticos t
- p-values
- Observaciones

## Regresion de ventas dado radio

In [18]:
X=np.hstack([x2])
lr=LinearRegression()
lr.fit(X,y)
lr.intercept_
lr.coef_
r2_score(y,lr.predict(X))
x_media=np.mean(x2)
y_media=np.mean(y)
n=len(y)
p=2
RSS=sum((lr.predict(x2)-y)**2)
sigma2=RSS/(n-p)
SE_Bo2=sigma2*(1/n + ((x_media**2)/sum((x2-x_media)**2)))
SE_Bo=np.sqrt(SE_Bo2)
SE_B12=sigma2/sum((x2-x_media)**2)
SE_B1=np.sqrt(SE_B12)
bo_bajo=lr.intercept_-2*SE_Bo
bo_alto=lr.intercept_+2*SE_Bo
b1_bajo=lr.coef_-2*SE_B1
b1_alto=lr.coef_+2*SE_B1
t_bo=lr.intercept_/SE_Bo
t_b1=lr.coef_/SE_B1
p_b0 = 2*(1 - stats.t.cdf(np.abs(t_bo), n-p))
p_b1 = 2*(1 - stats.t.cdf(np.abs(t_b1), n-p))
print('Regresion de ventas dado radio')
print('----------------------------------')
print(f'Los coeficientes son: b0={lr.intercept_}, b1={lr.coef_}')
print(f'El R^2 es: {r2_score(y,lr.predict(X))}')
print(f'Los errores estándar son: SE_b0={SE_Bo}, SE_b1={SE_B1}')
print(f'Los intervalos de confianza son: b0 entre({bo_bajo}, {bo_alto}), b1 entre({b1_bajo}, {b1_alto})')
print(f'Los valores t son: t_b0={t_bo}, t_b1={t_b1}')
print(f'El p-value para b0 es: {p_b0}')
print(f'El p-value para b1 es: {p_b1}')

Regresion de ventas dado radio
----------------------------------
Los coeficientes son: b0=9.311638095158283, b1=[0.20249578]
El R^2 es: 0.33203245544529525
Los errores estándar son: SE_b0=[0.5629005], SE_b1=[0.02041131]
Los intervalos de confianza son: b0 entre([8.1858371], [10.43743909]), b1 entre([0.16167317], [0.2433184])
Los valores t son: t_b0=[16.54224531], t_b1=[9.92076547]
El p-value para b0 es: [0.]
El p-value para b1 es: [0.]


Segun nuestros p-values ambos bo y b1 son diferentes de cero. Lo que queire decir que por cada dolar que invierta en radio me dara aproximadamente 0.2024 de ventas. Si no invierto en radio aun asi tendria 9.311 de ventas.

In [19]:
X=np.hstack([x3])
lr=LinearRegression()
lr.fit(X,y)
lr.intercept_
lr.coef_
r2_score(y,lr.predict(X))
x_media=np.mean(x3)
y_media=np.mean(y)
n=len(y)
p=2
RSS=sum((lr.predict(x3)-y)**2)
sigma2=RSS/(n-p)
SE_Bo2=sigma2*(1/n + ((x_media**2)/sum((x3-x_media)**2)))
SE_Bo=np.sqrt(SE_Bo2)
SE_B12=sigma2/sum((x3-x_media)**2)
SE_B1=np.sqrt(SE_B12)
bo_bajo=lr.intercept_-2*SE_Bo
bo_alto=lr.intercept_+2*SE_Bo
b1_bajo=lr.coef_-2*SE_B1
b1_alto=lr.coef_+2*SE_B1
t_bo=lr.intercept_/SE_Bo
t_b1=lr.coef_/SE_B1
p_b0 = 2*(1 - stats.t.cdf(np.abs(t_bo), n-p))
p_b1 = 2*(1 - stats.t.cdf(np.abs(t_b1), n-p))
print('Regresion de ventas dado periódico')
print('----------------------------------')
print(f'Los coeficientes son: b0={lr.intercept_}, b1={lr.coef_}')
print(f'El R^2 es: {r2_score(y,lr.predict(X))}')
print(f'Los errores estándar son: SE_b0={SE_Bo}, SE_b1={SE_B1}')
print(f'Los intervalos de confianza son: b0 entre({bo_bajo}, {bo_alto}), b1 entre({b1_bajo}, {b1_alto})')
print(f'Los valores t son: t_b0={t_bo}, t_b1={t_b1}')
print(f'El p-value para b0 es: {p_b0}')
print(f'El p-value para b1 es: {p_b1}')

Regresion de ventas dado periódico
----------------------------------
Los coeficientes son: b0=12.35140706927816, b1=[0.0546931]
El R^2 es: 0.05212044544430516
Los errores estándar son: SE_b0=[0.62142019], SE_b1=[0.01657572]
Los intervalos de confianza son: b0 entre([11.10856669], [13.59424744]), b1 entre([0.02154165], [0.08784454])
Los valores t son: t_b0=[19.87609562], t_b1=[3.29959074]
El p-value para b0 es: [0.]
El p-value para b1 es: [0.0011482]


De manera similar al pasado nuestra prueba de hipotesis nos indica que nuestros coeficientes son diferentes a cero. Si no meto nada a periodico obtengo 12.35 de ventas y cada vez que el periodico aumneta en una unidad gano 0.05469.

## Regresión lineal múltiple

En lugar de hacer una regresión para cada factor independiente, quizás se puede extender el modelo para que tenga varios factores dentro:

$$ Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_p X_p + \epsilon $$

Para nuestro ejemplo de publicidad:

$$ \text{sales} = \beta_0 + \beta_1 (\text{TV}) + \beta_2 (\text{radio}) + \beta_3 (\text{newspaper}) + \epsilon $$

Utiliza la librería `statsmodels` para realizar la regresión. Por defecto la librería statsmodels no toma en cuenta el intercepto ($\beta_0$), por lo que se tendrá que agregar una columna de unos de tamaño *n* a la matriz X.

`import statsmodels.api as sm`

`ols = sm.OLS(Y, X)`

`results = ols.fit()`

`results.summary()`

In [21]:
import statsmodels.api as sm

In [24]:
x=np.ones([n,1])
X=np.hstack([x, x1,x2,x3])

In [25]:
ols=sm.OLS(y,X)
results=ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Thu, 28 Aug 2025   Prob (F-statistic):           1.58e-96
Time:                        16:55:45   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
x1             0.0458      0.001     32.809      0.000       0.043       0.049
x2             0.1885      0.009     21.893      0.000       0.172       0.206
x3            -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

¿Qué diferencias puedes observar entre los *p-values* de una regresión múltiple y los encontrados en las regresiones simples? ¿Por qué crees que existen estas diferencias?

Puedo observar que el p-value de periódico es muy grande, lo que nos indica que no existe una relación entre las ventas y el periódico. Creo que existen estas diferencias, ya que ahora tenemos información más completa, ya que ahora consideramos todas las variables al mismo tiempo y la información que nos da el periódico ya se ve reflejada en las demás variables.

## Referencia

James, G., Witten, D., Hastie, T., Tibshirani, R.,, Taylor, J. (2023). An Introduction to Statistical Learning with Applications in Python. Cham: Springer. ISBN: 978-3-031-38746-3

## Addendum

Para calcular los *p-values* de los parámetros sin `statsmodels`:
1. Calcular RSS
2. Calcular RSE
3. `var_beta = np.linalg.inv(X.T @ X) * rse**2` (X con columna de intercepto)
4. `std_beta = np.sqrt(var_beta.diagonal())` El orden de los valores corresponde al orden de los factores en las columnas de la matriz $X$.
5. Calcular *estadístico t*
6. Calcular *p-value*